src/main/java/org/pac4j/vertx/handler/impl/CallbackHandler.java

```diff
@@ -51,75 +43,37 @@
     private final HttpActionAdapter httpActionHandler = new DefaultHttpActionAdapter();
     private final Vertx vertx;
     private final Config config;
+    private final boolean multiProfile;
 
     protected String defaultUrl = Pac4jConstants.DEFAULT_URL_VALUE;
+    private final CallbackLogic<Void, VertxWebContext> callbackLogic = new VertxCallbackLogic();
 
     public CallbackHandler(final Vertx vertx,
-                           final Config config) {
+                           final Config config,
+                           final Pac4jAuthHandlerOptions authHandlerOptions) {
         this.vertx = vertx;
         this.config = config;
+        this.multiProfile = authHandlerOptions.multiProfile();
     }
 
     @Override
     public void handle(RoutingContext event) {
 
         // Can we complete the authentication process here?
         final VertxWebContext webContext = new VertxWebContext(event);
-        final ProfileManager profileManager = new VertxProfileManager(webContext);
-        final Client client = config.getClients().findClient(webContext);
-
-        CommonHelper.assertNotNull("client", client);
-        CommonHelper.assertTrue(client instanceof IndirectClient, "only indirect clients are allowed on the callback url");
 
-        // Annoyingly because client.getCredentials can force a blocking wrap of a non-blocking call (eg when using clustered shared data)
-        // we have to wrap this in an executeBlocking call so that the future will be got on a non-eventloop thread!
-        vertx.<CommonProfile>executeBlocking(future -> {
-            try {
-                final Credentials credentials = client.getCredentials(webContext);
-                final CommonProfile profile = (CommonProfile) client.getUserProfile(credentials, webContext);
-                future.complete(profile);
-            } catch (final RequiresHttpAction e) {
-                LOG.warn("Requires http action: " + e.getCode());
-                httpActionHandler.handle(e.getCode(), webContext);
-                future.fail(e);
-            }
-        }, false, // Don't want to enforce ordering for this
-        result -> {
-            if (result.succeeded()) {
-                Optional.ofNullable(result.result()).ifPresent(commonProfile -> {
-                    profileManager.save(true, commonProfile);
+        vertx.executeBlocking(future ->
```

Comment:
It's the least imperfect wrapping :) The ideal wrapping would rely on a non-blocking pac4j which can offer callbacks (and use of non-blocking http clients implemented by the calling framework) but I think that's still a way off, as we'd have to define a non-blocking API and also a way of wrapping non-blocking http clients, for example.